src/ALCProxy.Communication/ICommObjects.cs

```diff
@@ -0,0 +1,257 @@
+        protected object _server;
+        protected string _serverTypeName;
+        protected Type _intType;
+        protected MethodInfo _callMethod;
+        public ALCClient(Type interfaceType, string serverName)
+        {
+            _intType = interfaceType;
+            _serverTypeName = serverName;
+        }
+        private Type FindTypeInAssembly(string typeName, Assembly a)
+        {
+            //find the type we're looking for
+            Type t = null;
+            foreach (Type ty in a.GetTypes())
+            {
+                if (ty.Name.Equals(typeName) || (ty.Name.StartsWith(typeName) && ty.Name.Contains("`"))) // will happen for non-generic types, generics we need to find the additional "`1" afterwards
+                {
+                    t = ty;
+                    break;
+                }
+            }
+            if (t == null)
+            {
+                //no type asked for in the assembly
+                throw new Exception("Proxy creation exception: No valid type while searching for the given type");
+            }
+            return t;
+        }
+        /// <summary>
+        /// Creates the link between the client and the server, while also passing in all the information to the server for setup
+        /// </summary>
+        /// <param name="alc">The target AssemblyLoadContext</param>
+        /// <param name="typeName">Name of the proxied type</param>
+        /// <param name="assemblyPath">path of the assembly to the type</param>
+        /// <param name="genericTypes">any generics that we need the proxy to work with</param>
+        public void SetUpServer(AssemblyLoadContext alc, string typeName, string assemblyPath, object[] constructorParams, Type[] genericTypes)
+        {
+            Assembly a = alc.LoadFromAssemblyPath(assemblyPath);
+            //find the type we're going to proxy inside the loaded assembly
+            Type objType = FindTypeInAssembly(typeName, a);
+            //Load *this* (The CommObjects) assembly into the ALC so we can get the server into the ALC
+            Assembly aa = alc.LoadFromAssemblyPath(Assembly.GetAssembly(typeof(ClientDispatch)).CodeBase.Substring(8));
+            Type serverType = FindTypeInAssembly(_serverTypeName, aa);
+            //Set up all the generics to allow for the serverDispatch to be created correctly
+            Type constructedType = serverType.MakeGenericType(_intType);
+            //Give the client its reference to the server
+            SerializeParameters(constructorParams, out IList<object> serializedConstArgs, out IList<Type> argTypes);
+            _server = constructedType.GetConstructor(
+                new Type[] { typeof(Type), typeof(Type[]), typeof(IList<object>), typeof(IList<Type>) })
+                .Invoke(new object[] { objType, genericTypes, serializedConstArgs.ToList(), argTypes });
+            _callMethod = _server.GetType().GetMethod("CallObject");
```

Comment:
I think we should be consistent about how we specify the server name and method - we shouldn't hard-code "CallObject" when the server type is passed into the ctor.